### ISSUE TITLE
NuttX Sample Monitor for ArduCopter

### ISSUE THREAD

**Issue by HenriqueMisson** [anmaped/rmtld3synth#3]
Hello Andre,

I followed all steps in the document to build the sample monitor for ardupilot, but I had an error during the compilation. Do you know what is wrong?

```
Pruning ROMFS files.
ROMFS:   romfs.img
OBJ:     romfs.o
CC:      romfs.o.c
CC:      romfs.o.c
LINK:    /home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.elf
/home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build//home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/builtin_commands.c.o:(.rodata.g_builtins+0xec): undefined reference to `rtmlib_app_main'
/home/labvant/ardupilot/modules/PX4Firmware/makefiles/firmware.mk:519: recipe for target '/home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.elf' failed
make[2]: *** [/home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.elf] Error 1
make[2]: Leaving directory '/home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build'
/home/labvant/ardupilot/modules/PX4Firmware/Makefile.make:143: recipe for target '/home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.px4' failed
make[1]: *** [/home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.px4] Error 2
make[1]: Leaving directory '/home/labvant/ardupilot'
../mk/px4_targets.mk:99: recipe for target 'px4-v2' failed
make: *** [px4-v2] Error 2
```

I'm using the arm-none-eabi 4.9.3

**Comment by anmaped** [comment]
` undefined reference to `rtmlib_app_main'` Could you check if the module is being compiled?

**Comment by HenriqueMisson** [comment]
Thank you very much Andre!

Now, i got it and I execute the monitor in nsh.

I got the monitor's message as : 

![rv_pixhawk](https://user-images.githubusercontent.com/37878221/48865045-7efa5a80-edb5-11e8-930c-55f4cf66ef30.png)

**Comment by HenriqueMisson** [comment]
I'm trying to analyze the results and I didn't understand why the RTML_monitor missed its deadline. The monitor doesn't observe the main task of the Copter?

**Comment by anmaped** [comment]
The period of the task is 0 `useconds: 0` and it has to be bigger in order to work properlly.

**Comment by HenriqueMisson** [comment]
Yes, I needed to instrument. I'm sorry for these type of questions, but I'm trying to understand your tool and I started to studying RV this year, so I don't have enough knowledge.

The best way to instrument the monitor in the ArduCopter code is insert directly  on the ```const AP_Scheduler::Task Copter::scheduler_tasks[]``` on "ArduCopter.cpp" or using the "UserCode.cpp" ?

**Comment by anmaped** [comment]
`AP_Scheduler` should be a good initial point.

**Comment by HenriqueMisson** [comment]
@anmaped , Did you add a file Ring_buffer.h in the rtmlib module for ardupilot? I didn't remember this file in the "original" rtmlib that is in your repository.

I tried to recompile ArduCopter again, and now I'm an error in this file, like 
 ```
Compiler executable checksum: 6948fa41b7fb2da910fed7adff42c5ba
In file included from /home/labvant/ardupilot/modules/rtmlib/RTML_writer.h:33:0,
                 from /home/labvant/ardupilot/modules/rtmlib/nuttx/main.cpp:13:
/home/labvant/ardupilot/modules/rtmlib/Ring_buffer.h:317:9: error: prototype for 'timeabs Ring_buffer<T>::getCounterCurrentTimestamp(uint64_t) const' does not match any in class 'Ring_buffer<T>'
 timeabs Ring_buffer<T>::getCounterCurrentTimestamp(uint64_t lcounter) const
         ^
compilation terminated due to -Wfatal-errors.
/home/labvant/ardupilot/modules/PX4Firmware/makefiles/module.mk:210: recipe for target 'main.cpp.o' failed
make[3]: *** [main.cpp.o] Error 1
make[3]: Leaving directory '/home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/home/labvant/ardupilot/modules/rtmlib/nuttx'
/home/labvant/ardupilot/modules/PX4Firmware/makefiles/firmware.mk:233: recipe for target '/home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build//home/labvant/ardupilot/modules/rtmlib/nuttx/module.pre.o' failed
make[2]: *** [/home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build//home/labvant/ardupilot/modules/rtmlib/nuttx/module.pre.o] Error 2
make[2]: Leaving directory '/home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build'
/home/labvant/ardupilot/modules/PX4Firmware/Makefile.make:143: recipe for target '/home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.px4' failed
make[1]: *** [/home/labvant/ardupilot/modules/PX4Firmware/Build/px4fmu-v2_APM.build/firmware.px4] Error 2
make[1]: Leaving directory '/home/labvant/ardupilot'
../mk/px4_targets.mk:99: recipe for target 'px4-v2' failed
make: *** [px4-v2] Error 2
```

**Comment by HenriqueMisson** [comment]
And do you know about the error:
 ```In file included from /home/labvant/ardupilot/modules/rtmlib/RTML_writer.h:33:0,
                 from /home/labvant/ardupilot/modules/rtmlib/nuttx/main.cpp:13:
/home/labvant/ardupilot/modules/rtmlib/Ring_buffer.h:317:9: error: prototype for 'timeabs Ring_buffer<T>::getCounterCurrentTimestamp(uint64_t) const' does not match any in class 'Ring_buffer<T>'
 timeabs Ring_buffer<T>::getCounterCurrentTimestamp(uint64_t lcounter) const
         ^
compilation terminated due to -Wfatal-errors.
```

**Comment by anmaped** [comment]
> 
> 
> And do you know about the error:
> 
> ```
>                 from /home/labvant/ardupilot/modules/rtmlib/nuttx/main.cpp:13:
> /home/labvant/ardupilot/modules/rtmlib/Ring_buffer.h:317:9: error: prototype for 'timeabs Ring_buffer<T>::getCounterCurrentTimestamp(uint64_t) const' does not match any in class 'Ring_buffer<T>'
> timeabs Ring_buffer<T>::getCounterCurrentTimestamp(uint64_t lcounter) const
>         ^
> compilation terminated due to -Wfatal-errors.
> ```
As I said It is fixed in 97f2c62.

**Comment by HenriqueMisson** [comment]
Ok, this was fixed. However now I have a 
```
/home/labvant/ardupilot/modules/rtmlib/nuttx/mon1/Rmtld3_reader.h:24:35: error: expected ')' before '<' token
       RMTLD3_reader(CircularBuffer<T> * const buffer, timespan ub) :      
                                   ^
compilation terminated due to -Wfatal-errors.
```
The CircularBuffer is now the Ring_Buffer? i tried this, but the error was the same. It looks like strange expected ')' before the '<'.

**Comment by anmaped** [comment]
You have to use the updated version of the rmtld3synt when synthesizing a monitor. I have performed several changes to it.

**Comment by HenriqueMisson** [comment]
I could recompile now! thanks

Now I'm seeing to instrument the monitor into the scheduler of ArduCopter.

**Comment by anmaped** [comment]
Yes. An old version of rmtld3synth, just compile the new one.

**Comment by anmaped** [comment]
I think I found a clue for you, so unpin dune first (`opam unpin dune`) and then install dune using ```opam pin add dune https://github.com/ocaml/dune/releases/download/1.5.0/dune-1.5.0.tbz```.

**Comment by anmaped** [comment]
Fixed 97f2c62.

**Comment by HenriqueMisson** [comment]
Hello André,

I tried to compile a monitor into Arducopter, and I got this error: ```error: ISO C++ forbids comparison between pointer and integer [-fpermissive]
         if ( getCounterCurrentPage(OLD_FRAME_ADDRESS) == &page )``` 

as you can see:
```
GNU C++ (15:4.9.3+svn231177-1) version 4.9.3 20150529 (prerelease) (arm-none-eabi)
	compiled by GNU C version 5.2.1 20151129, GMP version 6.1.0, MPFR version 3.1.3, MPC version 1.0.3
warning: MPFR header version 3.1.3 differs from library version 3.1.4.
GGC heuristics: --param ggc-min-expand=100 --param ggc-min-heapsize=131072
Compiler executable checksum: 6948fa41b7fb2da910fed7adff42c5ba
In file included from /home/labvant/ardupilot/modules/rtmlib/RTML_writer.h:33:0,
                 from /home/labvant/ardupilot/modules/rtmlib/nuttx/main.cpp:13:
/home/labvant/ardupilot/modules/rtmlib/Ring_buffer.h: In instantiation of 'void Ring_buffer<T>::enqueue(const T&, Ring_buffer<T>::tm_page&) [with T = int; Ring_buffer<T>::tm_page = Ring_buffer<int>::_page]':
/home/labvant/ardupilot/modules/rtmlib/RTML_writer.h:100:5:   required from 'void RTML_writer<T>::enqueue(const T&) [with T = int]'
/home/labvant/ardupilot/modules/rtmlib/nuttx/main.cpp:86:22:   required from here
/home/labvant/ardupilot/modules/rtmlib/Ring_buffer.h:387:55: error: ISO C++ forbids comparison between pointer and integer [-fpermissive]
         if ( getCounterCurrentPage(OLD_FRAME_ADDRESS) == &page )
                                                       ^
compilation terminated due to -Wfatal-errors.
/home/labvant/ardupilot/modules/PX4Firmware/makefiles/module.mk:210: recipe for target 'main.cpp.o' failed
make[3]: *** [main.cpp.o] Error 1
```
Do you know what can be?

PS: I tried to put the simple quotes, but another error appeared.

**Comment by anmaped** [comment]
Replace it by `getCounterCurrentPage(OLD_FRAME_ADDRESS) == (NATIVE_POINTER_TYPE) &page`. It should solve the problem.

**Comment by HenriqueMisson** [comment]
It worked! Thanks André.

After I had another error in the line ``` setCounterCurrentPage(new_value,  ntp);``` as
```
error: invalid conversion from 'Ring_buffer<int>::tm_page* {aka Ring_buffer<int>::_page*}' to 'uint32_t {aka unsigned int}' [-fpermissive]
         setCounterCurrentPage(new_value, ntp);
```
and I did the same think that you said in the previous answer and it worked fine.

**Comment by anmaped** [comment]
Set ` setCounterCurrentPage(new_value, (NATIVE_POINTER_TYPE) ntp);`.

**Comment by anmaped** [comment]
Fixed e603df1a713c753717ba0e0120d835d92563cda4.
